### DISCUSSION TITLE
How to K8SRunLauncher locally?

### DISCUSSION THREAD

**Discussion by amarrella** [dagster-io/dagster#3474] [Q&A]
Hi, I'm trying to run my pipeline using local dagit and K8SRunLauncher on an existing cluster but the job coordinator seems to fail.

I have set up my (local) instance like this:
```yaml
run_launcher:
    module: dagster_k8s.launcher
    class: K8sRunLauncher
    config:
        service_account_name: dagster
        dagster_home: "/opt/dagster/dagster_home"
        job_image:
            env: JOB_IMAGE
        instance_config_map: dagster-instance
        postgres_password_secret: dagster-postgresql-secret
        job_namespace: dagster
        load_incluster_config: False
        env_config_maps:
        - dagster-my-pipeline-user-env # this exists in the cluster
```

When I run the job, it successfully creates a pod, but the pod immediately fails with an error:
```yaml
Traceback (most recent call last):
  File "/usr/local/bin/dagster", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/site-packages/dagster/cli/__init__.py", line 41, in main
    cli(auto_envvar_prefix=ENV_PREFIX)  # pylint:disable=E1123
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.7/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/dagster/cli/api.py", line 93, in execute_run_with_structured_logs_command
    _execute_run_command_body(recon_pipeline, args.pipeline_run_id, instance, send_to_buffer)
  File "/usr/local/lib/python3.7/site-packages/dagster/cli/api.py", line 111, in _execute_run_command_body
    EngineEventData.in_process(pid, marker_end="cli_api_subprocess_init"),
  File "/usr/local/lib/python3.7/site-packages/dagster/core/instance/__init__.py", line 1019, in report_engine_event
    check.inst_param(pipeline_run, "pipeline_run", PipelineRun)
  File "/usr/local/lib/python3.7/site-packages/dagster/check/__init__.py", line 185, in inst_param
    obj, ttype, param_name, additional_message=additional_message
  File "/usr/local/lib/python3.7/site-packages/future/utils/__init__.py", line 446, in raise_with_traceback
    raise exc.with_traceback(traceback)
dagster.check.ParameterCheckError: Param "pipeline_run" is not a PipelineRun. Got None which is type <class 'NoneType'>.
```

When I describe the pods, it seems it's passing my local environment as a python environment (see the args below)
```
api
execute_run_with_structured_logs
{"__class__": "ExecuteRunArgs", "instance_ref": null, "pipeline_origin": {"__class__": "PipelinePythonOrigin", "pipeline_name": 
"my_pipeline", "repository_origin": {"__class__": "RepositoryPythonOrigin", "code_pointer": {"__class__": "ModuleCodePointer",
 "fn_name": "repository", "module": "bin.repository"}, "executable_path": "/Users/amarrella/Projects/my_project/.venv/bin/python"}}, 
"pipeline_run_id": "d69e4fbc-28a3-4ee5-89b1-5b8ff6c6fcd3"}
```

I'm not sure this is related but it's suspicious that it uses my local executable path.

I checked:
- The dagster guide for workspaces: https://docs.dagster.io/overview/repositories-workspaces/workspaces#main and tried to change the executable_path there, but the catch is that local dagit then won't work.
- The K8SRunLauncher API Doc: https://docs.dagster.io/_apidocs/libraries/dagster_k8s#dagster_k8s.K8sRunLauncher

To me all seems configured correctly, am I missing something?

I'm able to workaround the issue by using the dagit instance in the kubernetes cluster, but that requires updating the pods with the grpc server as well and i'd like to avoid this during fast iterations.

**Comment by catherinewu** [comment]
Hi @amarrella 👋 My suspicion is that the K8s run pod is not using the same run storage as your local instance; when launching a run, dagit's instance (in your case, the local instance) writes the pipeline run to the run storage, but the K8s pod might have a different run storage configured due to a different run storage on its instance config (resulting in None). 

To check, could you inspect the instance of the k8s job (`instance_config_map: dagster-instance`) and see what run storage is used? And then can also manually check that the expected run_id exists there (ie "d69e4fbc-28a3-4ee5-89b1-5b8ff6c6fcd3" from above)

**Reply by catherinewu** [reply]
I think running `kubectl describe configmaps dagster-instance` should work

**Reply by amarrella** [reply]
Hii @catherinewu 👋  Thanks for your reply! 

I see, it makes sense. My local dagit and my remote dagit are using a different DB as their RunStorage. My local one is using the default (a local SQLite) while the remote one is using a postgres DB hosted on k8s. I'll try pointing my local dagit to the remote DB and see what happens. 

Thanks again, and happy holidays to you and the rest of the Dagster team!

**Reply by catherinewu** [reply]
Hi @amarrella awesome -- let us know if you run into other issues. happy holidays to you too!

**Reply by amarrella** [reply]
Thanks again :) I got it to trigger correctly by setting up the run launcher as such and tunneling postgres:
```yaml
run_storage:
  module: dagster_postgres.run_storage
  class: PostgresRunStorage
  config:
    postgres_db:
      username: [redacted]
      password: [redacted]
      hostname: localhost
      db_name:  test
      port: 5432

```
it seems to work, but it also seems to ignore my k8s tags (to specify resource requirements & co). Is that a limitation of the K8SRunLauncher (as opposed to celery-k8s)?

**Reply by catherinewu** [reply]
Would you mind sharing the tag that the K8sRunLauncher is ignoring? The K8sRunLauncher should treat tags the same way as the CeleryK8sRunLauncher

**Reply by amarrella** [reply]
Hi @catherinewu, thanks for the reply. I think there is something else going on.

I was actually looking at the `dagster-run` pod but not the `dagster-job` but it's fine that the tags are not in the run. 

The problem is that the `dagster-job` is never fired and the job also seems stuck in the "pending" state in dagit. Meanwhile the dagster-run pod completes after a few seconds without doing anything.

The last line of the log in the run is 
```
{"__class__": "ExecuteRunArgsLoadComplete"}
```
so it seems it's not attempting to do anything after that.

**Reply by catherinewu** [reply]
Hi @amarrella!

Just to clarify -- when using the K8sRunLauncher, only a `dagster-run` job (+ pod) is created and not a `dagster-job` job (+ pod) so it's that part seems expected. 

Does the `dagster-run` pod contain any logs? do you see errors if you run `kubectl describe job dagster-run-...`?

**Reply by catherinewu** [reply]
the line right after logging `ExecuteRunArgsLoadComplete` is `pipeline_run = instance.get_run_by_id(pipeline_run_id)` so I think it's possible the pipeline_run_id wasn't found. We do have error handling set up for this so there should be an error in the pod logs if that's true

**Reply by amarrella** [reply]
Hi @catherinewu , apologies for the late reply! I retried today with a dummy pipeline like so:
```python
@dagster.solid(
    tags={
        "dagster-k8s/config": {
            "container_config": {
                "resources": {"requests": {"cpu": 1}, "limits": {"cpu": 1}}
            }
        }
    }
)
def test_tags(context: dagster.SolidExecutionContext) -> None:
    context.log.info("started")
    time.sleep(300)
    context.log.info("ended")


@dagster.pipeline
def test_pipeline() -> None:
    _ = test_tags()

```

The pipeline successfully runs but there are a few issues:
- on dagit, i only see the success, i don't see the logs (might be another thing i need to set in dagster.yaml perhaps)
- the problem with the tags still persists. When describing the `dagster-run-...` pod, there is no indication of the resource requests/limits speciified above.

**Reply by catherinewu** [reply]
Hi @amarrella I'm sorry I haven't been able to look into this yet. Will try to do so this week or ask someone else on the team to look. In the tests, we do assert that these requirements are successfully passed through the the k8s pod for the k8s run launcher, so not sure without tinkering around more

**Reply by catherinewu** [reply]
by "don't see the logs", does that refer to the events in the pipeline run view? do the events show up when refreshing the page after run succeeds? does this consistently happen or sporadically

**Reply by johannkm** [reply]
> When describing the dagster-run-... pod, there is no indication of the resource requests/limits speciified above.

Hi @amarrella - With the dagster-k8s run launcher we create a single pod for the entire pipeline run, so config can't be specified per solid. Instead you can place the tags on the pipeline definition with something like
```
  @pipeline(
      tags={
          "dagster-k8s/config": {
              "container_config": {
                  "resources": {
                      "requests": {"cpu": "250m", "memory": "64Mi"},
                      "limits": {"cpu": "500m", "memory": "2560Mi"},
                  }
              }
          }
      },
  )
  def foobar():
      ...
```
They can also be edited in the Dagit playground (though it's currently a bit messy).

**Reply by amarrella** [reply]
Hi @catherinewu @johannkm 👋 

thanks for bearing with me!

> by "don't see the logs", does that refer to the events in the pipeline run view? do the events show up when refreshing the page after run succeeds? does this consistently happen or sporadically

I don't see the events/logs i emit with `context.log` and that doesn't happen even after refresh. 

> Hi @amarrella - With the dagster-k8s run launcher we create a single pod for the entire pipeline run, so config can't be specified per solid. Instead you can place the tags on the pipeline definition with something like

Ah that's a pity. I'm stuck with celery so, as I don't want to over-provision resources (especially if they are GPU nodes).

Thanks again for all your help!

**Reply by catherinewu** [reply]
Hmm, could you try with `context.log.info`? I'm surprised that you don't see an error, hmm

**Reply by amarrella** [reply]
> Hmm, could you try with context.log.info? I'm surprised that you don't see an error, hmm

Hey @catherinewu , sorry i meant `context.log.info` not `context.log`. Unfortunately i can't see them.

Anyway, given that i can't really use this launcher for what i'm trying to do (i need different pods per solid and my understanding now is that the K8sRunLauncher triggers a pod per pipeline) I think i'll have use the CeleryK8sRunLauncher instead.

**Reply by amarrella** [reply]
I'm an idiot, I realized now why the run_coordinator was stuck... My docker image entrypoint in the image i was trying to run from localhost was starting dagit and ignoring everything else (the "live" image didn't have this issue). Removing the entrypoint fixed the issue

**Reply by catherinewu** [reply]
ohhhh i see, so glad that you found the issue!
